core/iwasm/compilation/aot_emit_variable.c

```diff
@@ -149,57 +150,156 @@ aot_compile_op_tee_local(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
+        LLVMBuildInBoundsGEP2(comp_ctx->builder, INT8_TYPE, import_global_data,
+                              &data_offset, 1, "import_global_value_ptr");
+
+    LLVMBuildBr(comp_ctx->builder, merge_block);
+
+    // Merge block
+    LLVMPositionBuilderAtEnd(comp_ctx->builder, merge_block);
+    LLVMValueRef phi =
+        LLVMBuildPhi(comp_ctx->builder, OPQ_PTR_TYPE, "global_value_addr");
+    LLVMAddIncoming(phi, &local_global_value_ptr, &then_block, 1);
+    LLVMAddIncoming(phi, &import_global_value_ptr, &else_block, 1);
+
+    return phi;
+}
+
 static bool
 compile_global(AOTCompContext *comp_ctx, AOTFuncContext *func_ctx,
                uint32 global_idx, bool is_set, bool is_aux_stack)
 {
     const AOTCompData *comp_data = comp_ctx->comp_data;
     uint32 import_global_count = comp_data->import_global_count;
-    uint32 global_base_offset;
-    uint32 global_offset;
     uint8 global_type;
-    LLVMValueRef offset, global_ptr, global, res;
+    LLVMValueRef global_ptr, global, res;
     LLVMTypeRef ptr_type = NULL;
 
-    global_base_offset = offsetof(AOTModuleInstance, global_table_data.bytes)
-                         + sizeof(AOTMemoryInstance)
-                               * (comp_ctx->comp_data->memory_count
-                                  + comp_ctx->comp_data->import_memory_count);
-
     bh_assert(global_idx < import_global_count + comp_data->global_count);
 
     if (global_idx < import_global_count) {
-        global_offset =
-            global_base_offset
-            /* Get global data offset according to target info */
-            + (comp_ctx->pointer_size == sizeof(uint64)
-                   ? comp_data->import_globals[global_idx].data_offset_64bit
-                   : comp_data->import_globals[global_idx].data_offset_32bit);
         global_type = comp_data->import_globals[global_idx].type.val_type;
     }
     else {
-        global_offset =
-            global_base_offset
-            /* Get global data offset according to target info */
-            + (comp_ctx->pointer_size == sizeof(uint64)
-                   ? comp_data->globals[global_idx - import_global_count]
-                         .data_offset_64bit
-                   : comp_data->globals[global_idx - import_global_count]
-                         .data_offset_32bit);
         global_type =
             comp_data->globals[global_idx - import_global_count].type.val_type;
     }
 
     if (comp_ctx->enable_gc && aot_is_type_gc_reftype(global_type))
         global_type = VALUE_TYPE_GC_REF;
 
-    offset = I32_CONST(global_offset);
-    if (!(global_ptr = LLVMBuildInBoundsGEP2(comp_ctx->builder, INT8_TYPE,
-                                             func_ctx->aot_inst, &offset, 1,
-                                             "global_ptr_tmp"))) {
-        aot_set_last_error("llvm build in bounds gep failed.");
-        return false;
-    }
+    global_ptr = get_global_value_addr(comp_ctx, func_ctx, global_idx);
```

Comment:
The original code directly gets the global offset at the compilation time and loads global one time in execution. Per my understanding, `get_global_value_addr` will get the global address by several loads, it will reduce the performance? It seems that it is only needed for import global and only needed when multi-module is enabled?

Reply:
OK, since performance and code size are important for the developers, hope the new feature impacts them as less as we can.